rusoto/services/s3/src/generated.rs

```diff
@@ -1028,6 +1028,12 @@ impl AnalyticsS3ExportFileFormatSerializer {
 
 pub struct StreamingBody(Box<Read>);
 
+impl StreamingBody {
+    pub fn new(read: Box<Read>) -> StreamingBody {
```

Comment:
Could we make this generic over a type parameter rather than requiring a `Box`, so that the signature is `fn new<R: Read>(read: R) -> StreamingBody`?

I feel like that would afford us more flexibility in the future without having to break the API.

Reply:
Apologies, I'm still learning Rust having a little difficulty making the edits you suggest:

If I make the signature more generic

```rust
impl StreamingBody {
    pub fn new<R: Read>(read: R) -> StreamingBody {
        StreamingBody(read)
    }
}
```

Then the compiler gets angry with  me:

```
error[E0308]: mismatched types
    --> services/s3/src/generated.rs:1033:24
     |
1033 |          StreamingBody(read)
     |                        ^^^^ expected struct `std::boxed::Box`, found type parameter
     |
     = note: expected type `std::boxed::Box<std::io::Read + 'static>`
                found type `R`
```

If I try to Box inside the constructor (which feels a little strange to me, hiding a heap allocation from the user)

```rust
impl StreamingBody {
    pub fn new<R: Read>(read: R) -> StreamingBody {
        StreamingBody(Box::new(read))
    }
}
```

This also throws an error:

```rust
error[E0310]: the parameter type `R` may not live long enough
    --> services/s3/src/generated.rs:1033:23
     |
1032 |     pub fn new<R: Read>(read: R) -> StreamingBody {
     |                -- help: consider adding an explicit lifetime bound `R: 'static`...
1033 |         StreamingBody(Box::new(read))
     |                       ^^^^^^^^^^^^^^
     |
```

Based on my experience with other languages I don't believe the compilers suggestion of adding 'static is the right solution here.

Thoughts?

Thanks!

Reply:
No problem at all! I think you were on the right track here, going with `Box::new` inside the constructor. Requiring `R: Read + 'static` to make that work seems okay to me.

What `R: 'static` means in this context is that the type that gets passed as `R` does not include any non-static lifetime parameters. This would be unsafe as `StreamingBody` does not track any lifetimes.

An alternative could be to parameterise it as `StreamingBody<'a>`, in which case we could require `R: Read + 'a`, however I feel like that would be making the API more complex for little gain.

Reply:
I briefly entertained what it would take to add the lifetime parameters to StreamingBody and you are correct it would certainly make things more complex, so went with the static approach instead.  I suppose if someday someone has a use case for passing in a reference to this constructor they could revisit this and look into adding lifetime parameters.